[SPARK-23367][Build] Include python document style checking

rekhajoshm opened the pull request:
## What changes were proposed in this pull request?
Includes python document style checking.
- Use sphinx like check, run only if pydocstyle installed on machine/jenkins
- use pydocstyle rather than single file version pep257.py, which  is much older and had some known issues
- verify pydocstyle latest 3.0.0  is in use, to ensure latest doc checks are getting executed
- ignore (inclusion/exclusion error codes) features and support via tox.ini
- Be non-breaking change and allow updating docstyle to standards at easy pace

## How was this patch tested?
./dev/run-tests

rekhajoshm committed:
Merge pull request #1 from apache/master

Pulling functionality from apache spark

rekhajoshm committed:
Merge pull request #2 from apache/master

pull latest from apache spark

rekhajoshm committed:
Merge pull request #3 from apache/master

Pulling functionality from apache spark

rekhajoshm committed:
Merge pull request #4 from apache/master

Pulling functionality from apache spark

rekhajoshm committed:
Merge pull request #5 from apache/master

pull request from apache/master

rekhajoshm committed:
Merge pull request #6 from apache/master

pull latest from apache spark

rekhajoshm committed:
Merge pull request #8 from apache/master

pull latest from apache spark

rekhajoshm committed:
Merge pull request #9 from apache/master

Pull apache spark

rekhajoshm committed:
Merge pull request #10 from apache/master

pull latest apache spark

rekhajoshm committed:
Merge pull request #11 from apache/master

Apache spark pull latest

rekhajoshm committed:
Merge pull request #12 from apache/master

Apache spark latest pull

rekhajoshm committed:
Merge pull request #13 from apache/master

apache spark latest pull

rekhajoshm committed:
Merge remote-tracking branch 'upstream/master'

rekhajoshm committed:
Merge pull request #14 from apache/master

latest apache spark

rekhajoshm committed:
Merge branch 'master' of https://github.com/rekhajoshm/spark

rekhajoshm committed:
Merge pull request #15 from apache/master

apache spark latest

rekhajoshm committed:
Merge branch 'master' of https://github.com/rekhajoshm/spark

rekhajoshm committed:
Merge pull request #16 from apache/master

Apache Spark master latest

rekhajoshm committed:
Merge branch 'master' of https://github.com/rekhajoshm/spark

rekhajoshm committed:
[SPARK-23367][Build] Include python document style checking

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/3121/
Test PASSed.

SparkQA commented:
**[Test build #96086 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96086/testReport)** for PR 22425 at commit [`af83d90`](https://github.com/apache/spark/commit/af83d906655a5ee5cef1806ab4805f39e6fc4da5).

SparkQA commented:
**[Test build #96086 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96086/testReport)** for PR 22425 at commit [`af83d90`](https://github.com/apache/spark/commit/af83d906655a5ee5cef1806ab4805f39e6fc4da5).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96086/
Test FAILed.

srowen commented on dev/tox.ini:
```diff
@@ -14,6 +14,8 @@
 
 [pycodestyle]
-ignore=E226,E241,E305,E402,E722,E731,E741,W503,W504
+ignore=E226,E241,E305,E402,E722,E731,E741,W503,W504,W605
```
Do we need to exclude 605? I thought we fixed all those...

rekhajoshm replied:
The intent of this PR was on pydocstyle changes, and this was pycodestyle break, so was unsure if that fix can be included in this PR. I had seen W605 code style broken in that build, and to make it pass added this. I agree completely the fix for W605 can be done with no-PR trivial fix., will check again.

holdenk replied:
I'm just confused why this would need to be changed in this PR -- hopefully just a hold over from the previous PR?

srowen commented on dev/tox.ini:
```diff
@@ -14,6 +14,8 @@
 max-line-length=100
 exclude=cloudpickle.py,heapq3.py,shared.py,python/docs/conf.py,work/*/*.py,python/.eggs/*,dist/*
+[pydocstyle]
+ignore=D100,D101,D102,D103,D104,D105,D106,D107,D200,D201,D202,D203,D204,D205,D206,D207,D208,D209,D210,D211,D212,D213,D214,D215,D300,D301,D302,D400,D401,D402,D403,D404,D405,D406,D407,D408,D409,D410,D411,D412,D413,D414
```
This is tiny, but is this blank line necessary? it seems to separate the heading from its content?

rekhajoshm replied:
Between [pydocstyle] to ignore is necessary and standard style. Post ignore the blank line was  added as per requested by last reviewer on previous PR - https://github.com/apache/spark/pull/20556#discussion_r168079061

holdenk replied:
I don't think that's what @ueshin was asking for, I think it was a blank line after the `ignore=...`, but if @ueshin is around we can see what @ueshin says. It's also relatively minor provided everything functions.

ueshin replied:
I just asked to add a line break at the end of file, and the current style looks good to me.

srowen replied:
Oh I was asking about the blank line at line 21. Isn't "ignore=" related to "[pydocstyle]"? compare to no blank line between 16-17. It's really trivial but hey.

srowen reviewed (commented):
I like the general idea, but this also turns off most checks? are there any that can be meaningfully re-enabled with a few fixes?

HyukjinKwon reviewed (commented):
I would also leave short explanations in PR description to justify why they should be skipped. Looks so many things are being skipped. FYI, this project does not yet claim PEP 257 as far as I know.

rekhajoshm commented:
Sorry, I was waiting from Feb2018 to Jul2018 on the previous PR  #20556  , and hence there is a bit of disconnect from this issue and delay in providing context in which this PR was created.

As per earlier discussions with prior reviewers, the intent of this PR was to make changes to enable pydocstyle but not to include all pydocstyle changes across all docs as that would need to be discussed later and be prioritized. I will relook at this again with your comments and check pydoc for style corrections.

srowen reviewed (commented):
Ok I think it's fine to focus on enabling this tool here and then making changes later as we unexclude warnings. 

I think we had separately fixed the w605 errors; might be worth checking if that's still needed.

holdenk reviewed (commented):
Thanks for reviving this PR I'm excited to get better style checkers integrated into Spark. My main concern is with the change to pycodestyle's list of excluded problems.

holdenk commented:
Gentle ping, whats up?

rekhajoshm committed:
updated for pydocstyle latest version, and for review comments

rekhajoshm commented:
Thank you @srowen @holdenk @HyukjinKwon for your review.

As per discussion, in this PR, we are enabling docstyle change but not updating pydocs. Hence tox.ini has doc error codes included.It can be taken off few at a time.

pydocstyle is newer correct way of handling doc style issues over pep257. It covers both pep257 and numpy convention style.  http://www.pydocstyle.org/en/3.0.0/error_codes.html

updated pydocstyle to latest 3.0.0.
removed W605 from tox.ini. (It had appeared in the build when this branch was created, not lingering from old PR.)

SparkQA commented:
**[Test build #97820 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97820/testReport)** for PR 22425 at commit [`cba18e8`](https://github.com/apache/spark/commit/cba18e8280d1f12c5a7d7961202df9097a77a0f1).

SparkQA commented:
**[Test build #97836 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97836/testReport)** for PR 22425 at commit [`cba18e8`](https://github.com/apache/spark/commit/cba18e8280d1f12c5a7d7961202df9097a77a0f1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4312/
Test PASSed.

srowen commented on dev/lint-python:
```diff
@@ -99,6 +104,29 @@ else
     echo "flake8 checks passed."
 fi
 
+# Check python document style, skip check if pydocstyle is not installed.
```
Nit: the indentation below looks different (4-space) from the rest of the script (2-space)

srowen replied:
Actually, seems like indentation is inconsistent in this file, so let's leave it for now

SparkQA commented:
**[Test build #97836 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97836/testReport)** for PR 22425 at commit [`cba18e8`](https://github.com/apache/spark/commit/cba18e8280d1f12c5a7d7961202df9097a77a0f1).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97836/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4349/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4361/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4368/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4370/
Test PASSed.

SparkQA commented:
**[Test build #98096 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98096/testReport)** for PR 22425 at commit [`cba18e8`](https://github.com/apache/spark/commit/cba18e8280d1f12c5a7d7961202df9097a77a0f1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4540/
Test PASSed.

SparkQA commented:
**[Test build #98096 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98096/testReport)** for PR 22425 at commit [`cba18e8`](https://github.com/apache/spark/commit/cba18e8280d1f12c5a7d7961202df9097a77a0f1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98096/
Test PASSed.
